{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "# gettting the range of the intensity and the number of pepetides\n",
    "import pandas as pd\n",
    "import os\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "direcroy_path = '/media/kusterlab/internal_projects/active/TOPAS/WP31/Playground/Retrospective_study/2023.06.22_AhS_PAPER_COHORT'\n",
    "transcripts_path = \"/media/kusterlab/internal_projects/active/TOPAS/WP31/FPKM/fpkms_230725mitINFORM.csv\"\n",
    "basket_annotation_path = \"/media/kusterlab/internal_projects/active/TOPAS/WP31/Playground/Retrospective_MTBs_Evaluation/TUPAC_SCORING_4th gen_230628.xlsx\"\n",
    "patietns_meta_data = '/media/kusterlab/internal_projects/active/TOPAS/WP31/Playground/Retrospective_MTBs_Evaluation/Metadata_Papercohort_230801.xlsx'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# intensity files\n",
    "PREPROCESSED_FP_INTENSITY = 'preprocessed_fp.csv'\n",
    "\n",
    "REGEX = r'(^\\S+-.+-\\S)|(Reporter intensity corrected)|(^P\\d{6}$)'\n",
    "Z_SCORES_PATH = 'full_proteome_measures_z.tsv'\n",
    "Z_SCORE_REGEX = r'(^\\S+-.+-\\S)|(Reporter intensity corrected)|(^P\\d{6}$) Z-score'\n",
    "BASKET_SCORES_PATH = 'basket_scores_4th_gen.tsv'\n",
    "\n",
    "REGEX_META = r'^Identification metadata'  # to get num identified pepetides in FP intensity file\n",
    "def replace_statement(x):\n",
    "    x = x.replace('num_peptides=','')\n",
    "    x = x.replace('|','')\n",
    "    x = x.replace(';','')\n",
    "    x = x.replace('detected in batch','0')\n",
    "    return x\n",
    "\n",
    "\n",
    "def unnest_proteingroups(df:pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Unnest the protein_groups A;B as two separate rows with the same values\n",
    "    the protein groups are the index of the the pandas dataframe df\n",
    "    \"\"\"\n",
    "    temp_df = df\n",
    "    temp_df['index'] = temp_df.index.str.split(';')\n",
    "    temp_df = temp_df.explode('index')\n",
    "    temp_df = temp_df.set_index('index')\n",
    "    return temp_df\n",
    "\n",
    "\n",
    "def get_corelation_scatter_plot_per_prtein(protein:str,transcripts_z_scores_df,proteomics_df,plot_saving_dir):\n",
    "    import seaborn as sns\n",
    "    import matplotlib.pyplot as plt\n",
    "    df1 = pd.DataFrame(transcripts_z_scores_df.loc[protein,:])\n",
    "    df2 = pd.DataFrame(proteomics_df.loc[protein,:])\n",
    "    df1.columns = ['transcripts']\n",
    "    df2.columns = ['proteomics']\n",
    "    all = pd.concat([df2,df1],axis=1)\n",
    "    sns.scatterplot(data=all, x=\"transcripts\", y=\"proteomics\")\n",
    "    plt.ylim(-4, 4) \n",
    "    plt.xlim(-4, 4)\n",
    "    plt.savefig(f'{plot_saving_dir}/{protein}.svg',dpi=300)\n",
    "\n",
    "\n",
    "def intersection(lst1, lst2):\n",
    "    return list(set(lst1) & set(lst2))\n",
    "\n",
    "\n",
    "def setdiff(lst1,lst2):\n",
    "    return(list(set(lst1) - set(lst2)))\n",
    "\n",
    "\n",
    "def get_correlation(protein:str,df1,df2):\n",
    "    ans = None\n",
    "    try:\n",
    "        df1 = pd.DataFrame(df1.loc[protein,:])\n",
    "        df2 = pd.DataFrame(df2.loc[protein,:])\n",
    "        all_df= pd.concat([df1,df2],axis=1)\n",
    "        all_df = all_df.dropna()\n",
    "        if len(all_df) > 20:\n",
    "            ans = all_df.corr().iloc[0,1]\n",
    "    except:\n",
    "        print(f'{protein} lead to error we put NA')\n",
    "        pass\n",
    "    return ans,len(all_df)\n",
    "\n",
    "\n",
    "\n",
    "def get_pep_number_from_protein_name(intensity_df:pd.DataFrame,protein_name:str,regex= r'^Identification metadata') -> pd.DataFrame:\n",
    "    \"\"\" \n",
    "    gets the number of the identified pepetides from the protein name across all the patients\n",
    "    :intensity_df: A pandas dataframe of the intensities with Identification metadata columns for each patient \n",
    "    :protein_name: the name of the protein\n",
    "    :USAGE :\n",
    "        get_pep_number_from_protein_name(fp_df,'EGFR')\n",
    "\n",
    "    \"\"\"\n",
    "    try:\n",
    "        #print(protein_name)\n",
    "        premeta_df = pd.DataFrame(intensity_df.loc[protein_name,:])\n",
    "        premeta_df = premeta_df.filter(regex=regex,axis=0)\n",
    "        premeta_df[protein_name] = premeta_df[protein_name].str.replace('num_peptides=|;','')\n",
    "        premeta_df[protein_name] = premeta_df[protein_name].str.replace('detected in batch','0') \n",
    "        premeta_df[protein_name] = pd.to_numeric(premeta_df[protein_name])\n",
    "        premeta_df['Sample name'] = premeta_df.index.str.replace('Identification metadata ','')\n",
    "        premeta_df.columns = ['num_pep','Sample name']\n",
    "        return premeta_df\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "\n",
    "\n",
    "def load_intensity_meta_data(instensitypath,key,regex= r'^Identification metadata' ):\n",
    "    print(f'reading {instensitypath}')\n",
    "    if os.path.exists(instensitypath):\n",
    "        cols = pd.read_csv(instensitypath,low_memory=False, nrows=10).filter(regex=regex).columns.tolist()\n",
    "        cols.append(key)\n",
    "        intensity_df = pd.read_csv(instensitypath,low_memory=False, usecols=cols)\n",
    "        intensity_df.index = intensity_df[key]\n",
    "        intensity_df = intensity_df.loc[:,~intensity_df.columns.duplicated()]\n",
    "        # intensity_df = intensity_df.fillna(0)\n",
    "        print('Intensity meta_data loaded')\n",
    "        return intensity_df.fillna('num_peptides=0;')\n",
    "    else:\n",
    "        return [f'{instensitypath} did not exist']\n",
    "\n",
    "\n",
    "\n",
    "def get_cv_from_melted_df(melted_df:pd.DataFrame,protein_col='Proteins_names',value_col='value') -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    calculated CV in a long format dataFrame\n",
    "    :melted_df: long_format data_Frame with at least two columns one for the protein names and one for the values\n",
    "    :protein_col: the column name of the proteins\n",
    "    :value_col: the column of the values\n",
    "    \"\"\"\n",
    "    cv_df = melted_df.groupby(protein_col)[value_col].agg(['std','mean'])\n",
    "    cv_df['cv'] = cv_df['std']/cv_df['mean']\n",
    "    cv_df['Gene names'] = cv_df.index\n",
    "    return cv_df\n",
    "\n",
    "\n",
    "\n",
    "def protein_remover_by_sparcity(df:pd.DataFrame,minimum_samples_inside=20) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Removes the sparcity based on the defined level from the dataframe\n",
    "    :df: dataframe with the proteins as the index and the samples as the columns\n",
    "    :minimum_samples_inside: the minimun number of samples without NA values that should exist not to exclude that protein\n",
    "    \n",
    "    \"\"\"\n",
    "    df = df.copy()\n",
    "    nan_count = len(df.columns) - df.isna().sum(axis = 1)\n",
    "    df.drop(index= nan_count[nan_count<minimum_samples_inside].index,inplace=True)\n",
    "    return df\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
